### PR TITLE
[14.0][IMP] oxigen_stock_barcodes_gs1: add 21 ia (serial number)

### PR DESCRIPTION
**Pull request by FrankC013**

### CODE REVIEW

File: oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py [hunk 1]
```diff
@@ -14,13 +14,31 @@ class WizStockBarcodesRead(models.AbstractModel):
 
     def process_lot(self, barcode_decoded):
```
Comment (eantones) [inline oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py]: Skip pre-commit error `# noqa: C901`
exemple: https://github.com/OCA/stock-logistics-barcode/blob/20bb317cc53aeaf9e33142ba069a654a5957cf4b/stock_barcodes/wizard/stock_barcodes_read_picking.py#L402

File: oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py [hunk 2]
```diff
@@ -14,13 +14,31 @@ class WizStockBarcodesRead(models.AbstractModel):
...
         res = super().process_lot(barcode_decoded)
+        if self.product_id.tracking == "serial":
+            lot_barcode = barcode_decoded.get("21", False)
+            lot = self.env["stock.production.lot"].search(
+                [("name", "=", lot_barcode), ("product_id", "=", self.product_id.id)]
```
Comment (eantones) [inline oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py]: add "company_id" but not using self.env.company, using the company of the current document, picking, inventory, etc..

File: oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py [hunk 3]
```diff
@@ -37,12 +55,25 @@ def process_barcode(self, barcode):
             # try the AI 240 'Additional product identification assigned
             # by the manufacturer'.
             product_barcode = barcode_decoded.get("240", False)
-        lot_barcode = barcode_decoded.get("10", False)
         product_qty = barcode_decoded.get("37", False)
         if product_barcode:
-            product = self.env["product.product"].search(
-                self._barcode_domain(product_barcode)
+            self.env.cr.execute(
```
Comment (eantones) [inline oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py]: Before the query, you should ensure that all fields involved are written to the database, add:
```python
# Flushing the relevant fields of the involved models
self.env['product.product'].flush(['barcode'])
self.env['product.template'].flush(['company_id'])
```

File: oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py [hunk 4]
```diff
@@ -37,12 +55,25 @@ def process_barcode(self, barcode):
...
+                WHERE LOWER(LPAD(pp.barcode, 14, '0')) = LOWER(%s)
+                AND (pt.company_id = %s OR pt.company_id IS NULL);
+                """,
+                (product_barcode, self.env.company.id),
             )
+            products = self.env.cr.fetchall()
+            product = self.env["product.product"].browse([x[0] for x in products])
+            if len(product) > 1:
+                self._set_messagge_info(
+                    "not_found", _("More than one product with this barcode")
```
Comment (eantones) [inline oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py]: Add to the message the `default_code` of the products with "the same" barcode

**Review by eantones — CHANGES_REQUESTED** [review]

File: oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py [hunk 5]
```diff
@@ -17,7 +17,11 @@ def process_lot(self, barcode_decoded):
         if self.product_id.tracking == "serial":
             lot_barcode = barcode_decoded.get("21", False)
             lot = self.env["stock.production.lot"].search(
-                [("name", "=", lot_barcode), ("product_id", "=", self.product_id.id)]
+                [
+                    ("name", "=", lot_barcode),
+                    ("product_id", "=", self.product_id.id),
+                    ("company_id", "=", self.env.company.id),
```
Comment (eantones) [inline oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py]: IT should be obtained from the document as inventory or picking not the user

File: oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py [hunk 6]
```diff
@@ -71,7 +78,9 @@ def process_barcode(self, barcode):
             product = self.env["product.product"].browse([x[0] for x in products])
             if len(product) > 1:
                 self._set_messagge_info(
-                    "not_found", _("More than one product with this barcode")
+                    "not_found",
+                    _("The next products have the same barcode: %s")
+                    % product.mapped("default_code"),
```
Comment (eantones) [inline oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py]: The `default_code` could be null, use the `barcode` for the message which we do know it has value and change the base message accordingly

**Review by eantones — CHANGES_REQUESTED** [review]

File: oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py [hunk 7]
```diff
@@ -16,11 +16,14 @@ def process_lot(self, barcode_decoded):
         res = super().process_lot(barcode_decoded)
         if self.product_id.tracking == "serial":
             lot_barcode = barcode_decoded.get("21", False)
+            operation = getattr(self, "picking_id", False) or getattr(
+                self, "inventory_id", False
+            )
```
Comment (eantones) [inline oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py]: What happens if neither of the operations exist? Should we return an error?

**Review by eantones — CHANGES_REQUESTED** [review]

**Review by eantones — APPROVED** [review]
LGTM!

**Review by eantones — APPROVED** [review]

File: oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py [hunk 8]
```diff
@@ -57,8 +110,20 @@ def process_barcode(self, barcode):
...
-            self.process_lot(barcode_decoded)
+            if self.product_id.tracking == "serial":
+                if not self.check_and_process_serial_lot(barcode_decoded):
+                    return False
+            else:
+                self.process_lot(barcode_decoded)
+            processed = True
+        if lot_barcode and self.product_id.tracking != "none":
+            if not self.process_lot(barcode_decoded):
+                return False
```
Comment (eantones) [inline oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py]:
```suggestion
```

File: oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py [hunk 9]
```diff
@@ -57,8 +110,20 @@ def process_barcode(self, barcode):
...
+        else:
+            lot_barcode = barcode_decoded.get("10", False)
         if lot_barcode and self.product_id.tracking != "none":
-            self.process_lot(barcode_decoded)
+            if self.product_id.tracking == "serial":
+                if not self.check_and_process_serial_lot(barcode_decoded):
+                    return False
+            else:
+                self.process_lot(barcode_decoded)
+            processed = True
```
Comment (eantones) [inline oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py]: Replace by
```python
        if self.product_id.tracking == "serial":
            if not self.process_serial_lot(barcode_decoded):
                return False
            processed = True
        elif self.product_id.tracking == "lot":
            lot_barcode = barcode_decoded.get("10", False)
            if lot_barcode:
                self.process_lot(barcode_decoded)
            processed = True

```

File: oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py [hunk 10]
```diff
@@ -13,14 +16,46 @@ class WizStockBarcodesRead(models.AbstractModel):
...
+                _(
+                    "The lot %s has been found but the reference %s does not match."
+                    % (lot_barcode, ref_barcode)
+                ),
+            )
+            return False
+        if not lot and self.option_group_id.create_lot:
+            lot = self._create_lot(barcode_decoded)
+        self.lot_id = lot
+        return True
```
Comment (eantones) [inline oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py]:
```suggestion
    def process_serial_lot(self, barcode_decoded):
        operation = getattr(self, "picking_id", False) or getattr(
            self, "inventory_id", False
        )
        if not operation:
            raise ValidationError(
                _(
                    "This record has inconsistent data. Delete the record and recreate it."
                )
            )
        lot_barcode = barcode_decoded.get("21", False)
        if lot_barcode:
            lot = self.env["stock.production.lot"].search(
                [
                    ("name", "=", lot_barcode),
                    ("product_id", "=", self.product_id.id),
                    ("company_id", "=", operation.company_id.id),
                ]
            )
            if lot:
                ref_barcode = barcode_decoded.get("10", False)
                if ref_barcode and lot.ref != ref_barcode:
                    self._set_messagge_info(
                        "not_found",
                        _(
                            "The lot %s has been found but the reference %s does not match."
                            % (lot_barcode, ref_barcode)
                        ),
                    )
                    return False
            else:
                if self.option_group_id.create_lot:
                    lot = self._create_lot(barcode_decoded)

            if lot:
                self.lot_id = lot

        return True
```

File: oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py [hunk 11]
```diff
@@ -13,14 +16,46 @@ class WizStockBarcodesRead(models.AbstractModel):
...
+            self._set_messagge_info(
+                "not_found",
+                _(
+                    "The lot %s has been found but the reference %s does not match."
+                    % (lot_barcode, ref_barcode)
+                ),
+            )
+            return False
+        if not lot and self.option_group_id.create_lot:
+            lot = self._create_lot(barcode_decoded)
```
Comment (eantones) [inline oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py]: Create new `_create_serial_lot`

**Review by eantones — CHANGES_REQUESTED** [review]

**Review by eantones — APPROVED** [review]
Good job!

File: oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py [hunk 12]
```diff
@@ -155,6 +155,9 @@ def process_barcode(self, barcode):  # noqa: C901
         if processed:
             if not self.check_option_required():
                 return False
+            if self.is_manual_confirm or self.manual_entry:
+                self._set_messagge_info("info", _("Review and confirm"))
+                return False
```
Comment (eantones) [inline oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py:155]: @FrankC013 Is this a bug introduced by us or it's from the original OCA module???

**Review by eantones — CHANGES_REQUESTED** [review]
Remember to put the "why" as a commit tittle, and it's "[bug](https://www.computerworld.com/article/2515435/moth-in-the-machine--debugging-the-origins-of--bug-.html)" not "back" 😉 
[FIX] oxigen_stock_barcodes_gs1: scan lines always confirmed

File: oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py (hunk 12 above)
Comment (FrankC013) [reply oxigen_stock_barcodes_gs1/wizard/stock_barcodes_read.py:155]: @eantones It is a bug introduced by us and also not handled by the original OCA module

**Review by eantones — APPROVED** [review]

**Review by eantones — APPROVED** [review]
Good job!

**Review by eantones — APPROVED** [review]

**Review by eantones — APPROVED** [review]

**Pull request merged**